larpandoracontent/LArControlFlow/TrackDirectionTool.cc

```diff
@@ -0,0 +1,1154 @@
+{
+    this->FitHitChargeVector(hitChargeVector1, fitResult1, numberHitsToConsider);
+    this->FitHitChargeVector(hitChargeVector2, fitResult2, numberHitsToConsider);
+}
+
+//----------------------------------------------------------------------------------------------------------------------------------
+
+void TrackDirectionTool::ComputeProbability(DirectionFitObject &fitResult)
+{
+  float deltaChiSquaredPerHit(fitResult.GetDeltaChiSquaredPerHit());
+
+  
+  if (deltaChiSquaredPerHit < 0.0 && deltaChiSquaredPerHit > -40.0) {
+    float beta = 0.01;
+    float alpha = 0.30465 + (0.00082051*fitResult.GetNHits()) - (0.12857*fitResult.GetMinChiSquaredPerHit());
+    if (alpha < 0) {
+      alpha = 0.000000000000000000000000000000001;
+    } 
+    float pmax = 0.90706 + (0.00011538*fitResult.GetNHits()) - (0.032143*fitResult.GetMinChiSquaredPerHit()); 
+    float x = std::abs(deltaChiSquaredPerHit);
+    float paa = (1.0/(2.0*alpha));
+    float pab = ((2.0*alpha*pmax)+(2.0*beta*pmax) - alpha - beta);
+    float pac = (pow(((alpha + beta)/beta), beta/alpha));
+    float p0 = 0.5 + paa*((pab)*(pac));
+    float pc = 0.5 + (p0 - 0.5)*(1.0 - exp(-alpha*x))*(exp(-beta*x));
+
+    fitResult.SetProbability(pc);
+  
+  }
+  else {
+    float probability(0.5);
+    fitResult.SetProbability(probability);
+    return;
+  }
+  
+}
+
+
+//---------------------------------------------------------------------------------------------------------------------------------
+
+  void TrackDirectionTool::PerformFits(HitChargeVector &hitChargeVector, HitChargeVector &forwardsFitPoints, HitChargeVector &backwardsFitPoints, int numberHitsToConsider, float &forwardsChiSquared, float &backwardsChiSquared, int &/*fitStatus1*/, int &/*fitStatus2*/)
+{
+
+  lar_content::TrackDirectionTool::LookupTable globalMuonLookupTable;
+  if (globalMuonLookupTable.GetMap().empty()){
+     globalMuonLookupTable.SetInitialEnergy(m_tableInitialEnergy); 
+     globalMuonLookupTable.SetBinWidth(m_tableStepSize);
+
+     FillLookupTable(globalMuonLookupTable, 105.7);
+  }
+
+  double TotalCharge = (0.f);
+  double TotalHitWidth = (0.f);
+  float trackLength(0.f);
+  
+   for (HitCharge &hitCharge : hitChargeVector)
+    {
+        TotalHitWidth += hitCharge.GetHitWidth();
+        TotalCharge += hitCharge.GetCharge();
+    }
+
+   this->GetTrackLength(hitChargeVector, trackLength);
+
+   lar_content::TrackDirectionTool::HitChargeVector binnedHitChargeVector;
+   BinHitChargeVector(hitChargeVector, binnedHitChargeVector);
+
+
+
+   //forwards fit
+   double particleMass(105.7);
+   double maxScale = 0;
+   if (trackLength != 0) {
+     maxScale = globalMuonLookupTable.GetMaxRange()/trackLength;
+     if (maxScale < 1.1) {
+       maxScale = 1.1;
+     }
+   }
+   LookupTable lookupTable = globalMuonLookupTable;
+
+   const int nParameters = 3;
```

Comment:
Is the block of code that follows essentially minuit?

And more generally, the fitting code here looks like a good candidate for its own function

Reply:
Yes, and done.